torch_xla/experimental/scan.py

```diff
@@ -82,14 +88,335 @@ def scan(fn, init, xs):
+
+  @pass_through_additional_inputs
+  def body_fn(num_iters: xb.Op, *args: xb.Op, additional_inputs: Tuple[xb.Op,
+                                                                       ...]):
+    carry, xs, fn_carry_history, ys = unpack(args)
+    xs_len_op = xb.Op.scalar(num_iters.builder(), xs_len, dtype=xb.Type.S64)
+    one = xb.Op.scalar(num_iters.builder(), 1, dtype=xb.Type.S64)
+    idx = xs_len_op - num_iters
+    x = [dynamic_slice(v, idx) for v in xs]
+    for i in range(len(carry)):
+      fn_carry_history[i] = dynamic_update_slice(fn_carry_history[i], carry[i],
+                                                 idx)
+    result = call_fn_computation(carry, x, additional_inputs)
+    for i in range(flat_carry_len):
+      carry[i] = result.get_tuple_element(i)
+    for i in range(flat_y_len):
+      y = result.get_tuple_element(i + flat_carry_len)
+      ys[i] = dynamic_update_slice(ys[i], y, idx)
+    return (num_iters - one,) + pack(carry, xs, fn_carry_history, ys)
+
+  res = xb.Op.mkwhile(builder.params(), cond_fn, body_fn)
+  computation = res.build('scan')
+
+  outputs = torch_xla._XLAC._xla_user_computation('xla::scan',
+                                                  builder.param_tensors(),
+                                                  computation)
+  # skip the last num_additional_inputs
+  outputs = outputs[:len(outputs) - num_additional_inputs]
+  # `1:` to skip `num_iters`
+  carry, xs, fn_carry_history, ys = unpack(outputs[1:])
+
+  # Unflatten tensors back to PyTrees
+  return tree_unflatten(carry, carry_spec), tree_unflatten(
+      fn_carry_history, carry_spec), tree_unflatten(ys, fn_output_y_spec)
+
+
+@pytreeify
+class Scan(torch.autograd.Function):
+
+  @staticmethod
+  def forward(ctx, fn, init, xs):
+    # Forward pass, save inputs for backward
+    ctx._fn = fn
+    with torch._C._AutoDispatchBelowAutograd():
+      carry, carry_history, ys = _scan_impl(fn, init, xs)
+    flat_carry_history, carry_spec = tree_flatten(carry_history)
+    flat_xs, xs_spec = tree_flatten(xs)
+    ctx.save_for_backward(*flat_carry_history, *flat_xs)
+    ctx._flat_carry_len = len(flat_carry_history)
+    ctx._carry_spec = carry_spec
+    ctx._xs_spec = xs_spec
+    return carry, carry_history, ys
+
+  @staticmethod
+  def backward(ctx, grad_carry, grad_carry_history, grad_ys):
+    fn = ctx._fn
+    flat_carry_len = ctx._flat_carry_len
+    carry_spec = ctx._carry_spec
+    xs_spec = ctx._xs_spec
+    tensors_list = ctx.saved_tensors
+    carry_history = tree_unflatten(tensors_list[:flat_carry_len], carry_spec)
+    xs = tree_unflatten(tensors_list[flat_carry_len:], xs_spec)
+
+    def detach_tensor(inp: torch.Tensor) -> torch.Tensor:
+      x = inp.clone().detach()
+      x.requires_grad = inp.requires_grad
+      return x
+
+    def step_fn(grad_carry, pytree: Tuple[torch.Tensor, torch.Tensor,
+                                          torch.Tensor]):
+      grad_y, carry, x = pytree
```

Comment:
is this a typo?

Reply:
I don't think so -- `pytree` is a tuple of the output grad at current step (`grad_y`), carry at the current step (`carry`), and input at current step (`x`)